New introductory notebook for model developers

cachafla opened the pull request:
## Internal Notes for Reviewers


## External Release Notes

New introductory notebook for model developers. This notebook gives developers a full introduction to the developer framework and teaches them how to document a model end to end. High level sections:

**1. Initialize the ValidMind Developer Framework**

ValidMind’s Developer Framework provides a rich collection of documentation tools and test suites, from documenting descriptions of datasets to validation and testing of models using a variety of open source testing frameworks.

**2. Start the model development process with raw data, run out-of-the box tests and add evidence to model documentation**

In this stage the notebook will provide details on how to access ValidMind's test repository of individual tests that you will use as building blocks to ensure a model is being built appropriately. The goal is to show how to run tests, investigate results and add tests results / evidence to the documentation.

For a full list of out-of-box tests please refer to: https://docs.validmind.com/guide/test-descriptions.html

**3. Learn how to implement custom tests**

Usually, model developers have a lot of their own custom tests and it is important to include this within the model documentation. We will show how you how to include custom tests and then how they can be added to documentation as additional evidence.

**4. Ensure completion of documentation**

In this stage the notebook will show how to ensure that model documentation includes custom tests and how to make test configuration changes that apply to all tests in the model documentation template. At the end of this section you should have a fully documented model ready for review.

AnilSorathiya commented on notebooks/tutorials/intro_for_model_developers.ipynb:
```diff
@@ -0,0 +1,1688 @@
+    "\n",
+    "To do this, go to the documentation page of your model and navigate to the `Data Preparation` -> `Correlations and Interactions` section. Then hover after the \"Pearson Correlation Matrix\" content block to reveal the `+` button as shown in the screenshot below.\n",
+    "\n",
+    "![screenshot showing insert button for test-driven blocks](../images/insert-test-driven-block.png)\n",
```
Just minor point here: It would be good if we crop this image specially `Precision Recall Curve` in the image. Here we are inserting text driven block in the `Data Preparation` section but it seems like we are showing `Model performance` tests below. 
<img width="1027" alt="image" src="https://github.com/validmind/developer-framework/assets/30263958/61788349-b6e6-447b-b522-78960a399503">

cachafla replied:
Fixed.

AnilSorathiya reviewed (approved):
Just left a minor comment on image otherwise LGTM. 
Pretty Cool 👍

johnwalz97 reviewed (approved):
This is a fantastic notebook! I really like how it iteratively builds up the concepts as the dataset and model are built and documented!

lgtm 👌

MichaelIngvarRoenning commented:
All in all an **excellent** notebook, very clear on how the model developers should use it. A few minor items:

1. "Verify & Preview the documentation template" - change ValidMnd to ValidMind (I know very minor)

2. At Section number 2 we are directly loading in a dataset from ValidMind, however we should very briefly state what the dataset is, what we are trying to predict so that it's not an unknown for the user. 

I was about to write more comments on adding individual tests etc. but then I got to the section where this occurred and from my perspective this is done really well. 

However, I will say that part "Using External Test Providers" is a little advanced for a lot of Model Developers. This is definitely for the Model Developer that is an advanced user of Python and Github so we need to be careful to assume they will understand the instructions there. For example, "inline test" terminology is not something model developers will do. I think it could be useful to make this section more complete with a "real" example showing screenshots of a repository of tests and then combing it with the description already in the notebook.

MichaelIngvarRoenning reviewed (commented):
All in all an excellent notebook, very clear on how the model developers should use it. A few minor items:

"Verify & Preview the documentation template" - change ValidMnd to ValidMind (I know very minor)

At Section number 2 we are directly loading in a dataset from ValidMind, however we should very briefly state what the dataset is, what we are trying to predict so that it's not an unknown for the user.

I was about to write more comments on adding individual tests etc. but then I got to the section where this occurred and from my perspective this is done really well.

However, I will say that part "Using External Test Providers" is a little advanced for a lot of Model Developers. This is definitely for the Model Developer that is an advanced user of Python and Github so we need to be careful to assume they will understand the instructions there. For example, "inline test" terminology is not something model developers will do. I think it could be useful to make this section more complete with a "real" example showing screenshots of a repository of tests and then combing it with the description already in the notebook.

juanmleng commented on notebooks/tutorials/intro_for_model_developers.ipynb:
```diff
@@ -0,0 +1,1688 @@
+   "source": [
+    "## Register a new model in ValidMind UI and initialize the client library\n",
+    "\n",
+    "ValidMind generates a unique _code snippet_ for each registered model to connect with your developer environment. You initialize the client library with this code snippet, which ensures that your documentation and tests are uploaded to the correct model when you run the notebook.\n",
```
Is the 'code snippet' associated with a specific 'model' or a 'project'? Since the current code snippet requires a 'project' ID, perhaps we should clarify this.

cachafla replied:
Yep, we need to add support for `model` as an argument to `vm.init`. We're still maintaining support for `project` in some places.

We'll tackle that with another PR in the future.

juanmleng commented on notebooks/images/high-pearson-correlation-block.png:
In the screenshot of the high correlation pairs, the comment says: “Age column is present in the top two pair correlation pairs”, but the figure shows the Age column only present in the bottom pair.

cachafla replied:
Fixed.

juanmleng commented on notebooks/tutorials/intro_for_model_developers.ipynb:
```diff
@@ -0,0 +1,1688 @@
+   },
+   "outputs": [],
+   "source": [
+    "TRAIN = X_train\n",
```
Pedantic comment: Not very important, but just flagging for consistency shake. Should we adhere to a specific naming convention for dataframes, such as using lowercase and appending `_df` to clarify that these are dataframes?

cachafla replied:
Pedantic comment accepted.

juanmleng commented on notebooks/tutorials/intro_for_model_developers.ipynb:
```diff
@@ -0,0 +1,1688 @@
+   "source": [
+    "## ValidMind at a glance\n",
+    "\n",
+    "ValidMind's platform enables organizations to identify, document, and manage model risks for all types of models, including AI/ML models, LLMs, and statistical models. As a model developer, you use the ValidMind Developer Framework to automate documentation and validation tests, and then use the ValidMind AI Risk Platform UI to collaborate on model documentation. Together, these products sim ...
```
Just checking if we are using 'ValidMind AI Risk Platform UI' as the official name for our UI. If so, could the name potentially confuse developers who aren't working with AI? The comment after that, referencing "model risk management" should help clarifying that tho.

cachafla replied:
Leaving this for a future update since it's part of a template block we use for other notebooks.

juanmleng reviewed (approved):
This is a fantastic notebook. The concepts are explained very clearly, and the examples really help in showcasing the features. I just left some cosmetic comments. Other than that LGTM!

nrichers commented:
@cachafla I made a very light editing pass with focus on:

- Intro and outro tweaks
- Heading levels fixes (e.g. one of the main sections was an H3 but should have been an H2)
- Consistency edits and tone & style tweaks

Also added a table of contents and updated the "About ValidMind" section. 

I would recommend a more thorough edit later on when we have more time.

cachafla commented:
> "Verify & Preview the documentation template" - change ValidMnd to ValidMind (I know very minor)

done

cachafla commented:
> At Section number 2 we are directly loading in a dataset from ValidMind, however we should very briefly state what the dataset is, what we are trying to predict so that it's not an unknown for the user.

done

cachafla commented:
> For example, "inline test" terminology is not something model developers will do

added a small sentence explaining that

cachafla commented:
> @cachafla I made a very light editing pass with focus on:
> 
> * Intro and outro tweaks
> * Heading levels fixes (e.g. one of the main sections was an H3 but should have been an H2)
> * Consistency edits and tone & style tweaks
> 
> Also added a table of contents and updated the "About ValidMind" section.
> 
> I would recommend a more thorough edit later on when we have more time.

Thanks a lot! 🙏
